src/Platforms/DB2111Platform.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace Doctrine\DBAL\Platforms;
+
+use Doctrine\DBAL\Exception;
+
+use function sprintf;
+
+/**
+ * Provides the behavior, features and SQL dialect of the IBM DB2 11.1 (11.1 GA) database platform.
+ *
+ * @see https://www.ibm.com/docs/en/db2/11.1?topic=database-whats-new-db2-version-111-ga
+ */
+class DB2111Platform extends DB2Platform
+{
+    /**
+     * {@inheritdoc}
+     *
+     * @see https://www.ibm.com/docs/en/db2/11.1?topic=subselect-fetch-clause
+     */
+    protected function doModifyLimitQuery($query, $limit, $offset)
+    {
+        if ($limit !== null) {
+            if ($limit < 0) {
+                throw new Exception(sprintf('Limit must be a positive integer or zero, %d given', $limit));
+            }
+
+            if ($offset === 0) {
+                // In cases where an offset isn't required, we can use "FETCH FIRST".
+                $query .= sprintf(' FETCH FIRST %u ROWS ONLY', $limit);
+            } else {
+                $query .= sprintf(' OFFSET %u ROWS FETCH NEXT %u ROWS ONLY', $offset, $limit);
+            }
+        } elseif ($offset > 0) {
+            // 9223372036854775807 is the maximum of unsigned BIGINT, the biggest possible limit.
+            $query .= sprintf(' OFFSET %1$u ROWS FETCH NEXT (9223372036854775807 - %1$u) ROWS ONLY', $offset);
```

Comment:
This looks oddly complicated. This is how the implementation on 4.0.x looks like:

https://github.com/doctrine/dbal/blob/d12da23c3b295911c5e7067321841976e772c01c/src/Platforms/DB2Platform.php#L510-L521

If that implementation works, I'd prefer it over yours. If it doesn't, please elaborate.

Reply:
Sure. Confirmed it works without `LIMIT` in DB2 :+1: